File: leetcode/101-200/122-best-time-to-buy-and-sell-stock-ii/bench/max_profit.c
/* Benchmark workload — greedy O(n) Best Time to Buy and Sell Stock II.
 * Algorithmic mirror of bench/max_profit.kara. See ../README.md § Benchmarks for N / K and the LCG.
 * The metal floor: a flat malloc'd long[]. */
#include <stdio.h>
#include <stdlib.h>

static long long max_profit(const long long *prices, long long n) {
    long long profit = 0;
    for (long long i = 1; i < n; i++) {
        long long d = prices[i] - prices[i - 1];
        if (d > 0) profit += d;
    }
    return profit;
}

int main(void) {
    long long n = 2000000;
    long long *data = (long long *)malloc(sizeof(long long) * n);
    long long state = 12345;
    for (long long i = 0; i < n; i++) {
        state = (state * 1103515245 + 12345) & 2147483647;
        data[i] = (state & 4095) + 1;
    }
    long long sum = 0;
    for (int k = 0; k < 10; k++) sum += max_profit(data, n);
    printf("%lld\n", sum);
    free(data);
    return 0;
}

File: leetcode/101-200/122-best-time-to-buy-and-sell-stock-ii/bench/max_profit.rs
//! Benchmark workload — greedy O(n) Best Time to Buy and Sell Stock II.
//!
//! Algorithmic mirror of bench/max_profit.kara and bench/max_profit.py. See
//! ../README.md § Benchmarks for the choice of N / K and the deterministic LCG
//! generator. Kara checks integer overflow by default, so the like-for-like row
//! is `rustc -O -C overflow-checks=on`.

fn max_profit(prices: &[i64]) -> i64 {
    let mut profit: i64 = 0;
    for i in 1..prices.len() {
        let d = prices[i] - prices[i - 1];
        if d > 0 {
            profit += d;
        }
    }
    profit
}

fn main() {
    const N: usize = 2_000_000;
    let mut data: Vec<i64> = Vec::with_capacity(N);
    let mut state: i64 = 12345;
    for _ in 0..N {
        state = (state.wrapping_mul(1103515245).wrapping_add(12345)) & 2_147_483_647;
        data.push((state & 4095) + 1);
    }

    let mut sum: i64 = 0;
    for _ in 0..10 {
        sum += max_profit(&data);
    }
    println!("{}", sum);
}
